asyncpg/connresource.py

```diff
@@ -5,17 +5,47 @@
 # This module is part of asyncpg and is released under
 # the Apache 2.0 License: http://www.apache.org/licenses/LICENSE-2.0
 
+from __future__ import annotations
 
 import functools
+from typing import TYPE_CHECKING, Generic, Protocol, TypeVar
+from typing_extensions import ParamSpec
 
 from . import exceptions
 
+if TYPE_CHECKING:
+    from . import connection
 
-def guarded(meth):
+_ConnectionResourceT = TypeVar(
+    "_ConnectionResourceT", bound="ConnectionResource", contravariant=True
```

Comment:
The `bound` needs to be stringified, even with `from __future__ import annotations`. I checked this locally.